### PR TITLE
Predicting for the test set

### PR DESCRIPTION
**Pull request by paulbsalamanca**
We have run into a problem with the final step for the project. When we went to run our model on the hold out data, it did not run. We initially thought this was because the X hold out data is not processed the same as the X training data. To resolve this, we replicated the processing of the X training data onto the hold out data. The problem we are running in to now is that there are only ~1000 features for the processed data after we dummify it, however we have over ~2000 features in the training data. We guess this is because there is fewer/different variation in the categorical variables of the test set because it has much fewer observations than the training set. Because of this we are not able to run the best model on the hold out data.

### CODE REVIEW

**Comment by paulbsalamanca** [comment]
The code for preprocessing and trying to predict on the test set is at the end of the notebook. We hardcoded the optimal model (logistic regression on upsampled data) so that you don't have to run the whole notebook before trying to test out that code.

**Comment by Akesari12** [comment]
One quick way to approach this would be to take your logit model that did best and retrain it with just the top 10 or 20 features (based on coefficient magnitude), then subset X_test based on the columns in that reduced model. Alternatively you can subset the original training data based on columns that do show up in test. In a policy context you'll generally be constrained by what's available in the data used for deployment so refitting the model to work within those bounds is the best way to go. Does this make sense? Let me know if not!

**Comment by paulbsalamanca** [comment]
Did a little bit more digging. It’s definitely a problem with the dummy variables. 

Specifically, the dummification of Facility_Type, WARD_PRECINCT, and LICENSE_DESCRIPTION are what’s tripping us up. Because X_test has different (and much fewer) observations, its values for those categorical variables are gonna be generally different and result in different dummifications.

**Comment by paulbsalamanca** [comment]
> One quick way to approach this would be to take your logit model that did best and retrain it with just the top 10 or 20 features (based on coefficient magnitude), then subset X_test based on the columns in that reduced model. Alternatively you can subset the original training data based on columns that do show up in test. In a policy context you'll generally be constrained by what's available in the data used for deployment so refitting the model to work within those bounds is the best way to go. Does this make sense? Let me know if not!

Thanks Aniket!! Totally makes sense.
